Show git diff signs in gutter

sudormrfbin opened the pull request:
Shows the changes to the current file symbolically in the gutter:

| Without diagnostics | With diagnostics |
| --- | --- |
| ![Screenshot_2022-02-07_00-06-27](https://user-images.githubusercontent.com/23398472/152696286-4b292447-791a-4a85-9038-0bfae7d7767c.png) | ![Screenshot_2022-02-07_00-09-14](https://user-images.githubusercontent.com/23398472/152696301-87c9a2eb-9a10-46a1-833c-bf78533a0ed6.png) |

It looks a bit too cramped with diagnostics, so maybe the reserved space should be increased but that will make the gutter permanently too big (there's currently no way to "toggle" a gutter component). Changes are reflected only when the file is saved. I would like to show an accurate diff even if the file is unsaved, but this should be a good starting point.

The symbols are hard coded for now to keep the PR simple, and a follow up PR will make it configurable. The default might also need changing (though I would like to avoid too much bike shedding ;)

Supersedes #467.

pickfire commented on helix-view/src/editor.rs:
```diff
@@ -268,6 +269,8 @@ impl Editor {
         config: Config,
     ) -> Self {
         let language_servers = helix_lsp::Registry::new();
+        let vcs =
+            helix_core::find_root(None, &[]).and_then(|p| helix_vcs::Git::discover_from_path(&p));
```
Why not just name it `version_control` so below we can do `version_control,`?

pickfire commented on helix-view/src/gutter.rs:
```diff
@@ -36,6 +36,35 @@ pub fn diagnostic<'doc>(
+    _width: usize,
+) -> GutterFn<'doc> {
+    Box::new(move |line: usize, _selected: bool, out: &mut String| {
+        let diff = match doc.line_diffs().get(&(line + 1)) {
+            None => return None,
+            Some(d) => d,
+        };
```
```suggestion
        let diff = doc.line_diffs().get(&(line + 1))?;
```

pickfire commented on helix-view/src/gutter.rs:
```diff
@@ -36,6 +36,35 @@ pub fn diagnostic<'doc>(
+
+        let (icon, style) = match diff {
+            helix_vcs::LineDiff::Added => ("│", Style::default().fg(crate::graphics::Color::Green)),
+            helix_vcs::LineDiff::Deleted => ("▸", Style::default().fg(crate::graphics::Color::Red)),
```
I think this is not obvious that the current line is being deleted.

archseer replied:
Agreed, let's use the symbols I proposed in the other PR

sudormrfbin replied:
I used the right aligned version of all the symbols to have enough space between the diagnostic symbols and git signs:

![Screenshot_2022-02-07_10-41-16](https://user-images.githubusercontent.com/23398472/152728056-153e9fce-6d7a-420c-bdd2-4572a59f740e.png)

archseer replied:
What if we place it left aligned in that 1 char buffer between the line number and text?

sudormrfbin replied:
I tried some variations of what you suggested:

| Left aligned thick bar | Right aligned thick bar | Center aligned thin bar |
| --- | --- | --- |
| ![thick bar left](https://user-images.githubusercontent.com/23398472/152744862-7a98122e-5844-459c-a743-6fea26d31b53.png) | ![thick bar right](https://user-images.githubusercontent.com/23398472/152744905-bf0b5f39-b3ca-42f6-9ac5-49cc54d6f80f.png) | ![thin bar normal](https://user-images.githubusercontent.com/23398472/152744936-eee735f7-3a63-4550-a95c-441bdd88c482.png) |

It looks weird when horizontally scrolling though, so I would suggest sticking with making it a normal gutter component:

![horizontalscrol](https://user-images.githubusercontent.com/23398472/152745331-8a25bee9-2f47-4ff8-af5b-a19839f21064.png)

archseer replied:
A thin bar could work. We can make the space between the text and the gutter a gutter component, it's just currently hardcoded. Then `gutters = [diagnostics, line_number, git]`

archseer replied:
You can try more blocks from https://en.wikipedia.org/wiki/Block_Elements#Character_table

"Left one quarter block" or "Left three eights" seem most suitable

archseer replied:
This would be sort of similar to how VSCode draws it ![](https://assets.digitalocean.com/articles/git-integration-in-visual-studio-code/7.png)

I like it because it uses the existing gap instead of growing the view gutter space by another character

sudormrfbin replied:
On the other hand different terminals handle box drawing characters differently, so for example while the centered thin line looks okay on kitty, it look too thin on wezterm.

archseer commented on helix-view/src/gutter.rs:
```diff
@@ -36,6 +36,32 @@ pub fn diagnostic<'doc>(
+        let diff = doc.line_diffs().get(&(line + 1))?;
+
+        let (icon, style) = match diff {
+            helix_vcs::LineDiff::Added => ("▐", Style::default().fg(crate::graphics::Color::Green)),
+            helix_vcs::LineDiff::Deleted => ("▗", Style::default().fg(crate::graphics::Color::Red)),
+            helix_vcs::LineDiff::DeletedAbove => {
+                ("▝", Style::default().fg(crate::graphics::Color::Red))
+            }
+            helix_vcs::LineDiff::Modified => {
+                ("▐", Style::default().fg(crate::graphics::Color::Yellow))
+            }
```
Fetch colors from the theme here:

https://github.com/helix-editor/helix/blob/23553bd37c5e936c022faa70bef346448a36d6a2/theme.toml#L38-L40

p-e-w commented:
Hi folks, I stumbled upon this PR by accident. As it happens, [I implemented the exact same feature for the Micro editor](https://github.com/zyedidia/micro/pull/1487) a while ago. Here's how it looks:

![](https://user-images.githubusercontent.com/2702526/74081755-866c5900-4a78-11ea-8aa9-f14275261bed.png)

Two things from Micro's implementation work quite nicely and might be worth considering for Helix as well:

1. The gutter uses the character `\u2594` (UPPER ONE EIGHTH BLOCK) to indicate deleted lines on the following line. Compared to the arrow or block glyphs from this PR, this makes it completely clear *where* code was removed, and also has the advantage of resembling a minus sign, something people already associate with removal.

2. Unlike most diff gutter implementations, Micro does not ask Git (or other VCSs) for the diff. Instead, it has the concept of a "diff base", which is the text from which to diff, and the diff is *computed by the editor.* This has many advantages, like being so fast you can update the diff gutter on every keystroke (for small to medium sized files), and being easy to adapt to other VCSs (all you need is a way to ask the VCS for the committed state of the file). It also allows cool features like diffing from the saved state for files that aren't even part of a VCS repository.

You can try out the implementation by installing and running Micro, pressing Ctrl+E and entering the command `set diffgutter on`. Then open any file (Ctrl+O) from a Git repository.

Anyway, that's all. I absolutely love what you are building here. Have a great day!

sudormrfbin commented:
Hi @p-e-w, thanks for reaching out ! The upper one eighth block looks perfect for deletion and I've replaced the previous symbol with it. Since the discussion on the the placement of diff gutter also kind of got stalled, I went with the left aligned blocks for the added and modified markers with the gutter itself placed to the left edge.

I also like the idea of doing the diffing in the editor itself, especially if it can make update-as-you-type fast enough to be implemented. Currently it's updated only on save, which makes for a sub-optimal experience IMO. We already pull in the [`similar`](https://github.com/mitsuhiko/similar) diffing library for computing diffs on reloading files from disk, so this wouldn't add any new dependency either. What do you think @archseer ?

archseer commented:
We can do the diffing, but you'd still need a way to query for the original file.

sudormrfbin commented:
The diff is now updated live, on any change. The main problem left to tackle is to bypass converting the document Rope to a string every time the diff needs to be updated (so practically on most keypresses).

sudormrfbin commented:
Diffing avoids string conversion now by wrapping a Rope, implemented in 0cd310bb 🎉

archseer commented on helix-vcs/src/rope.rs:
```diff
@@ -0,0 +1,20 @@
+
+impl<'a> RopeLine<'a> {
+    pub fn from_rope(rope: &'a Rope) -> Vec<Self> {
+        rope.lines().into_iter().map(RopeLine).collect()
```
Note: `.lines()` is not very efficient: https://github.com/cessen/ropey/issues/25

sudormrfbin replied:
Ah I read `O(1)` in the chunks iterator and assumed lines would also follow suit. @cessen how hard/big of a change would it be to make the lines iterator also O(1) ? I suppose if it was easy enough it would've already been added.

sudormrfbin replied:
From casual use it seems snappy and I haven't noticed any performance issues, so this is probably not too serious of an issue :)

cessen replied:
> But then again this is a hotpath since it's called on almost every keystroke

Synchronously?

Thinking about it more, I would expect the diffing itself to be expensive enough that if the `Lines` iterator is causing performance issues then the diffing would be as well (i.e. even if `Lines` was infinitely fast, you'd still have issues).  So if this feature does turn out to cause performance problems (e.g. on large files or with large diffs), then rearchitecting it to work in a separate thread or something like that might be needed anyway.

archseer commented on helix-vcs/src/git.rs:
```diff
@@ -0,0 +1,197 @@
+        let timeout = Duration::from_millis(250);
+        let diff = similar::capture_diff_slices_deadline(
+            similar::Algorithm::Myers,
+            &old,
+            &new,
```
Couldn't you use rope's `chunks()` here rather than lines?

sudormrfbin replied:
I'm not sure whether two different ropes will always give chunks that are 1:1 with each other, and even if they did, we will have to condense the chunk diff to a line diff by querying the rope again for the line index of each diff chunk.

cessen commented on helix-vcs/src/git.rs:
```diff
@@ -0,0 +1,197 @@
+        };
+
+        let timeout = Duration::from_millis(250);
+        let diff = similar::capture_diff_slices_deadline(
```
If I'm reading your code correctly, it looks like you're collecting all the lines into a `Vec` in order to pass them to `capture_diff_slices_deadline()`, so even though this has a deadline set you can still be bottle necked on collecting everything into a `Vec`.

Instead, you could use `capture_diff_deadline()` by wrapping the `Rope` in a type that implements `Index` and returns the lines through indexing.  That way the entire procedure can be deadlined.

I could also make this even easier for you by implementing `Hash` on `RopeSlice` (which is really just an accidental omission on my part in Ropey).

sudormrfbin replied:
We only diff using `similar` initially and now use changesets to incrementally compute the diffs, so it's no longer executed on every keypress. It'll still be nice to optimize it using indexing since we don't want to stall on huge text files on startup.

archseer commented:
Thinking about this some more, it seems to me that we could compute changed ranges based on helix changesets (after maybe doing an initial diff against git?). We already have all the information needed to incrementally update the diff in the transaction rather than having to go through git and diffing full files every single time (these files could also be very large).

cessen commented on helix-vcs/src/rope.rs:
```diff
@@ -0,0 +1,20 @@
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub struct RopeLine<'a>(pub ropey::RopeSlice<'a>);
+
+impl Hash for RopeLine<'_> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        for chunk in self.0.chunks() {
+            chunk.hash(state);
+        }
+    }
+}
```
I've now implemented `Hash` directly on `Rope` and `RopeSlice` (in the recently released Ropey 1.4.0), so this should be unnecessary now.

It's also worth noting that your implementation here is incorrect (not to criticize, just an interesting technical note), which I only noticed when diving into the trait to implement it myself:

`&str`'s `Hash` implementation, which you're calling in the loop here, is purposefully implemented so that hashing a series of string slices yields a different hash than hashing the same data as a single concatenated string.  Which means that if two `Rope`s have identical contents, but the chunks happen to be split up differently, your implementation will give two different hashes for them.

Instead, you actually need to call `state.write(chunk.as_bytes())` in the inner loop.

The reason `&str`'s `Hash` implementation works like that is because of `#[derive(Hash)]`.  Imagine this struct:

```rust
#[derive(Hash)]
struct Person<'a> {
    first_name: &'a str,
    second_name: &'a str,
}
```

If concatenated strings gave the same hash, then someone named "Jon Athanbor" and "Jonathan Bor" would has to the same value if the names were stored in lower case.  (A contrived example, I know, but it's just to illustrate.)

archseer replied:
Yeah usually you prefix each chunk with a hash of the `len` to solve for this

sudormrfbin replied:
Ah, I didn't know this; thanks for pointing it out ! Thanks for the Hash impl too :)

poliorcetics commented:
> Thinking about this some more, it seems to me that we could compute changed ranges based on helix changesets (after maybe doing an initial diff against git?). We already have all the information needed to incrementally update the diff in the transaction rather than having to go through git and diffing full files every single time (these files could also be very large).

What about changes at a distance, like adding a "use std::path::Path;" with rust-analyzer from a use point hundreds of line down the line ? Is the information for those available ?

archseer commented:
Yes, all changes are always applied through a changeset/transaction.

poliorcetics commented:
Nice !

sudormrfbin commented on helix-view/src/document.rs:
```diff
@@ -648,12 +704,32 @@ impl Document {
 
         if !transaction.changes().is_empty() {
             self.version += 1;
+            let reverted_tx = transaction.invert(&old_doc);
+
+            // Consider H to be the text of the file in git HEAD, and B₁ to be
+            // the text when buffer initially loads. H → B₁ is the changeset
+            // that describes the diff between HEAD and buffer text. Inverting
+            // this produces B₁ → H, which is initially saved to `diff_changes`.
+            // In the next edit, buffer text changes to B₂. The transaction
+            // describes the change B₁ → B₂. Inverting this gives B₂ → B₁.
+            // Composing this with the saved `diff_changes` gives us
+            // (B₂ → B₁) → (B₁ → H) = B₂ → H. Whenever examining a change X₁ → X₂,
+            // we need to know the contents of the text at state X₁ to know where
+            // to apply the operations in the changeset. The advantage of inverting and
+            // composing this way instead of simply composing (which would give
+            // us H → B₂ in this example) is that we no longer need the HEAD text
+            // and we can instead use the current text in the buffer.
+            if let Some(changes) = self.diff_changes.take() {
+                let reverted_changes = reverted_tx.changes().clone();
+                let changes = reverted_changes.compose(changes);
+                self.diff_changes = Some(changes);
+                self.diff_with_base();
+            }
```
We use changesets now, and use the approach described in the comment. It also has the advantage that we don't have to manage a cache of the file contents in HEAD since we don't need it to generate the diff.

This works fine for most of the cases except for deleting something and undoing it (there are some other rough corners, but they can be easily solved at the rendering level). Normally, composing a change and it's undo operation produces an empty changeset. But since we use inverted changes here, it doesn't apply cleanly.

I'll write a better explanation of the problem tomorrow, but it boils down to the fact that composing for example `[Delete(5)]` and `[Insert("hello")]` (note the order) does not produce an empty changeset, where as composing an insert and _then_ a delete gives a changeset of `[]`.

Relevant test case:
<details><summary>Patch for the test</summary>

```diff
:100644 100644 daf4a77e 00000000 M	helix-core/src/transaction.rs

diff --git a/helix-core/src/transaction.rs b/helix-core/src/transaction.rs
index daf4a77e..9aed26e8 100644
--- a/helix-core/src/transaction.rs
+++ b/helix-core/src/transaction.rs
@@ -579,6 +579,17 @@ mod test {
     use super::*;
     use crate::State;
 
+    #[test]
+    fn composing_delete_and_insert_leaves_empty_changeset() {
+        let rope = Rope::from("");
+        let mut delete = ChangeSet::new(&rope);
+        delete.delete(5);
+        let mut insert = ChangeSet::new(&rope);
+        insert.insert("hello".into());
+
+        let changes = delete.compose(insert);
+        assert_eq!(changes.changes, &[]);
+    }
     #[test]
     fn composition() {
         use Operation::*;
```
</details>

```rust
fn composing_delete_and_insert_leaves_empty_changeset() {
    let rope = Rope::from("");
    let mut delete = ChangeSet::new(&rope);
    delete.delete(5);
    let mut insert = ChangeSet::new(&rope);
    insert.insert("hello".into());

    let changes = delete.compose(insert);
    assert_eq!(changes.changes, &[]);
}
```

pickfire commented on helix-vcs/Cargo.toml:
```diff
@@ -0,0 +1,19 @@
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+git2 = { version = "0.13", default-features = false }
```
I wonder if we really want to use the c bindings here instead of something that is written in rust?

pickfire commented on helix-vcs/src/lib.rs:
```diff
@@ -0,0 +1,49 @@
+
+#[derive(Debug, Default)]
+pub struct Registry {
+    inner: HashMap<RepoRoot, Rc<RefCell<Git>>>,
```
Not sure if it's a good idea to have this as `Git` given that it is called vcs, I thought it is supposed to be extensible to other vcs.

pickfire commented on helix-view/src/document.rs:
```diff
@@ -612,6 +624,50 @@ impl Document {
         self.language_server = language_server;
     }
 
+    pub fn set_version_control(&mut self, vcs: Option<Rc<RefCell<helix_vcs::Git>>>) {
+        self.version_control = vcs;
+    }
```
Why not just do `version_control_mut` for this?

pickfire commented on helix-view/src/document.rs:
```diff
@@ -612,6 +624,50 @@ impl Document {
+            let changes = compare_ropes(diff_base, self.text()).changes().clone();
+            let changes = changes.invert(diff_base);
+            self.diff_changes = Some(changes);
+            drop(vcs);
```
Can you please leave a comment on the drop here?

univerz commented:
it would be great if this could be easily toggled - it's a useful feature, but it's visually distracting to see it all the time.

sudormrfbin commented:
Closing in favor of #3890. Thanks @pascalkuthe for picking it up!
